### ISSUE TITLE
[naming-convention] variable or variableLike which is an object

### ISSUE THREAD

**Issue by ehossack** [typescript-eslint/typescript-eslint#2149]
**What were you trying to do?**

```javascript
{
  "rules": {
                    "@typescript-eslint/naming-convention": [
                    "error",
                    { "selector": "variable", 'types': ['boolean', 'string', 'number', 'array'], "format": ["camelCase", "UPPER_CASE"], leadingUnderscore: 'forbid' },
                    { "selector": "variable", 'types': ['function'], "format": ["camelCase", "PascalCase"], leadingUnderscore: 'forbid' },
                    // this section
                    { "selector": "variable", 'types': ['object'], "format": ["PascalCase"], leadingUnderscore: 'forbid' },
  }
}
```

```ts
export const ModuleOpenApiComponents: OpenAPIV3.ComponentsObject = {
// ....
};
```

**What did you expect to happen?**
I expected that there was a way to be able to specify `PascalCase` for object types. Or at the very least the rest types, if I were to do something like:
```json
{ "selector": "variableLike", "format": [ "PascalCase"], "leadingUnderscore": "forbid" },
```
after all those rules, the 'fallback' result for everything not specified under the types of the first two variable rules would be to allow `PascalCase`

**What actually happened?**
```
 error  Variable name ModuleOpenApiComponents must match one of the following formats: camelCase, UPPER_CASE  @typescript-eslint/naming-convention
```

Possibly there does exist a way to specify what I'd like, but I'm just not able to discern it from the documentation, in which case I suggest we improve the docs :)

**Versions**

| package                                 | version |
| --------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`      | `3.1.0` |
| `@typescript-eslint/parser`             | `3.1.0` |
| `TypeScript`                            | `3.9.3` |
| `eslint`                            | `7.1.0` |
| `node`                                  | `13.10.0` |
| `npm`                                   | `6.13.7` |

**Comment by bradzacher** [comment]
The main driver for the `types` selector is to allow users to:
- require boolean variables start with a verb like `isFoo`, `hasBar`, etc
- require array variables are pluralised `foos`, `lollies`, etc

Could you explain your use case a bit more?
Why do you want to specifically name objects with upper-case?

**Comment by ehossack** [comment]
Gotcha. For us, our codebase has "modules" which are typed via interfaces to have properties like components, or functions, e.g.
```
interface MyCodebaseModule {
    plugin: SomeLibraryPlugin;
    components?: OpenApiV3.ComponentsObject;
    docs?: DocumentationObject;
}
```

And then we would export that object as such:
```
export const TheHealthCheckModule: MyCodebaseModule = {
  plugin: (app) => app.register('/healthcheck', (req) => 'OK')),
}
```

Thinking about it from your statement, we could define rules for how our primitives, arrays and functions look, and say everything else could be `camelCase` or `PascalCase`.
I can confirm that this works with the following:
```
                    { "selector": "variable", 'types': ['boolean', 'string', 'number', 'array'], "format": ["camelCase", "UPPER_CASE"], leadingUnderscore: 'forbid' },
                    { "selector": "variable", 'types': ['function'], "format": ["camelCase", "PascalCase"], leadingUnderscore: 'forbid' },
                    { "selector": "function", "format": ["camelCase", "PascalCase"], leadingUnderscore: 'forbid' },
                    { "selector": "variable", "format": ["camelCase", "PascalCase"], leadingUnderscore: 'forbid' },
```

So maybe in this case, the docs could be updated simply around the behaviour of this rule when it comes to object types?

**Comment by bradzacher** [comment]
We can add `object` as a type. It's just not a use case I'd really thought of.
Happy to accept a PR

**Comment by karlhorky** [comment]
Aside from `object`, it would also be great to specify what type the values are in the object, eg. via `Record<string, function>` or similar.

**Comment by karlhorky** [comment]
Another use case for either `object` or `Record<string, x>`:

`react-navigation`'s `Stack` object (an object of functions which are React components). This is a pattern in other React libraries too.

```ts
import { createNativeStackNavigator } from '@react-navigation/native-stack';

const Stack = createNativeStackNavigator();

function MyStack() {
  return (
    <Stack.Navigator>
      <Stack.Screen name="Home" component={Home} />
      <Stack.Screen name="Notifications" component={Notifications} />
      <Stack.Screen name="Profile" component={Profile} />
      <Stack.Screen name="Settings" component={Settings} />
    </Stack.Navigator>
  );
}
```

**Comment by alecmev** [comment]
Would also be nice to be able to tell apart objects and instances too. In our case, we want to enforce `UPPER_CASE` on all top-level constants except functions and instances, but **not** objects or any other types.

**Comment by JoshuaKGoldberg** [comment]
Per #8792, we're feature freezing this rule. It's reached the point of complexity where we're having a hard time justifying working on it. 

If you'd like to see this option happen, I'd encourage you to add it to a community project such as [eslint.style](https://eslint.style). Cheers! 💙

**Comment by karlhorky** [comment]
I also mentioned this just now over in my question to the ESLint Stylistic maintainers about potentially migrating this rule to ESLint Stylistic:

- https://github.com/typescript-eslint/typescript-eslint/issues/8792#issuecomment-2025344673
